xrootd/xrdfuse/xrdfuse.go

```diff
@@ -0,0 +1,166 @@
+// Copyright 2018 The go-hep Authors.  All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package xrdfuse contains implementation of a FUSE api for the XRootD.
+package xrdfuse
+
+import (
+	"context"
+	"log"
+	"os"
+	"path"
+
+	"github.com/hanwen/go-fuse/fuse"
+	"github.com/hanwen/go-fuse/fuse/nodefs"
+	"github.com/hanwen/go-fuse/fuse/pathfs"
+	"go-hep.org/x/hep/xrootd/client"
+	"go-hep.org/x/hep/xrootd/xrdfs"
+	"go-hep.org/x/hep/xrootd/xrdproto"
+)
+
+type fs struct {
+	pathfs.FileSystem
+	client *client.Client
+	xrdfs  xrdfs.FileSystem
+	root   string
+}
+
+// NewFS returns a new path.FileSystem representing filesystem on the remote XRootD server.
+// client is a client to the remote XRootD server.
+// root is the path to the remote directory to be used as a root directory.
+func NewFS(client *client.Client, root string) pathfs.FileSystem {
+	return &fs{
+		FileSystem: pathfs.NewDefaultFileSystem(),
+		client:     client,
+		xrdfs:      client.FS(),
+		root:       root,
+	}
+}
+
+// GetAttr implements pathfs.FileSystem.GetAttr
+func (fs *fs) GetAttr(name string, ctx *fuse.Context) (*fuse.Attr, fuse.Status) {
+	stat, err := fs.xrdfs.Stat(context.Background(), path.Join(fs.root, name))
+	if serverError, ok := err.(xrdproto.ServerError); ok {
+		if serverError.Code == 3011 {
+			// File does not exists.
+			return nil, fuse.ENOENT
+		}
+	}
+	if err != nil {
+		log.Printf("xrdfuse: error calling stat: %v\n", err)
+		return nil, fuse.EIO
+	}
+
+	var out fuse.Attr
+	out.Size = uint64(stat.Size())
+	out.Mtime = uint64(stat.Mtime)
+	out.Mode = entryStatToMode(stat)
+
+	return &out, fuse.OK
+}
+
+// OpenDir implements pathfs.FileSystem.OpenDir
+func (fs *fs) OpenDir(name string, ctx *fuse.Context) (c []fuse.DirEntry, code fuse.Status) {
+	entries, err := fs.xrdfs.Dirlist(context.Background(), path.Join(fs.root, name))
+	if err != nil {
+		log.Printf("xrdfuse: error calling dirlist: %v\n", err)
```

Comment:
perhaps, instead of logging this and hoping somebody might look at the screen, we could store this `err` value on the side ? (as a field of `fs` ?)

Reply:
What about having a `chan error`, so we will be able to receive all errors in `xrootd-fuse` and log them?

Reply:
that could be an option.
we should have a look at how it could be used in real life as a user, though.

Reply:
It looks like something like `FS.AddErrorHandler(...)` will be better since it will avoid possible blocks and will not require extra cleanup.

What do you think about it?